File: examples/csharp/Scripts/DemoSceneBase.cs
using Godot;

namespace FairyGUI.Examples;

public partial class DemoSceneBase : Node
{
    protected const string MainMenuScenePath = "res://csharp/Scenes/MainMenu.tscn";

    protected GRoot _groot;
    protected GuiObject _view;

    private bool _sceneActive;
    private string _pendingScenePath = "";

    public override void _Ready()
    {
        _sceneActive = true;
        RegisterDefaultFonts();
        if (GRoot.GetInstance() != null)
            GRoot.Cleanup();
        CallDeferred(MethodName.DelayedInit);
    }

    public override void _ExitTree()
    {
        _view = null;
        _groot = null;
        _sceneActive = false;
        _pendingScenePath = "";
    }

    protected bool IsSceneActive() => _sceneActive;

    protected bool IsUiActive()
    {
        if (!_sceneActive || !IsInsideTree())
            return false;
        if (_groot == null || _view == null)
            return false;
        if (GRoot.GetInstance() == null)
            return false;
        return GRoot.GetInstance() != null;
    }

    protected SceneTree GetEngineTree()
    {
        return Engine.GetMainLoop() as SceneTree;
    }

    protected SceneTree SafeGetTree()
    {
        if (!_sceneActive || !IsInsideTree())
            return null;
        return GetTree();
    }

    private void DelayedInit()
    {
        if (!IsSceneActive())
            return;
        var tree = SafeGetTree();
        if (tree == null)
            return;
        if (GRoot.GetInstance() == null)
            GRoot.Create(tree);
        DeferredAttachToGroot();
    }

    private void DeferredAttachToGroot()
    {
        if (!IsSceneActive())
            return;
        _groot = GRoot.GetInstance();
        PrepareGrootForScene();
        CallDeferred(MethodName.FinishContinueInit);
    }

    private void FinishContinueInit()
    {
        if (!IsSceneActive() || _groot == null)
            return;
        ContinueInit();
        AddCloseButton();
    }

    private void PrepareGrootForScene()
    {
        if (_groot == null)
            return;
        if (DragDropManagerHelper.GetInstance().IsDragging())
            DragDropManagerHelper.GetInstance().Cancel();
        _groot.HideTooltips();
        _groot.HidePopup();
        _groot.CloseModalWait();
        _groot.CloseAllWindows();
        _groot.RemoveChildren();
    }

    private void RegisterDefaultFonts()
    {
        const string cjkFont = "res://Resources/fonts/DroidSansFallback.ttf";
        const string latinFont = "res://Resources/fonts/arial.ttf";
        var ui = UIConfigHelper.GetInstance();
        ui.RegisterFont("default", cjkFont);
        ui.RegisterFont("en", latinFont);
        ui.RegisterFont("微软雅黑", cjkFont);
        ui.RegisterFont("Consolas", latinFont);
        ui.RegisterFont("Comic Sans MS", latinFont);
        ui.RegisterFont("Arial", latinFont);
        ui.SetDefaultFont("default");
    }

    protected void UseLatinDefaultFontOnWeb()
    {
        if (!Engine.HasFeature("web"))
            return;
        UIConfigHelper.GetInstance().SetDefaultFont("en");
    }

    protected void RestoreCjkDefaultFont()
    {
        UIConfigHelper.GetInstance().SetDefaultFont("default");
    }

    protected virtual void ContinueInit()
    {
    }

    protected async Task WaitSeconds(double seconds)
    {
        if (!IsSceneActive())
            return;
        var tree = SafeGetTree() ?? GetEngineTree();
        if (tree == null)
            return;
        await ToSignal(tree.CreateTimer(seconds), SceneTreeTimer.SignalName.Timeout);
    }

    private void AddCloseButton()
    {
        var closeBtn = UIPackage.CreateObject("MainMenu", "CloseButton");
        if (closeBtn == null)
            return;
        closeBtn.SetPosition(_groot.GetWidth() - closeBtn.GetWidth() - 10, _groot.GetHeight() - closeBtn.GetHeight() - 10);
        closeBtn.AddRelation(_groot, GuiObject.RelationType.RightRight, false);
        closeBtn.AddRelation(_groot, GuiObject.RelationType.BottomBottom, false);
        closeBtn.SetSortingOrder(100000);
        closeBtn.AddClickListener(new Callable(this, MethodName.OnClose));
        _groot.AddChild(closeBtn);
    }

    protected void CleanupGrootOverlays()
    {
        if (_groot == null)
            return;
        if (DragDropManagerHelper.GetInstance().IsDragging())
            DragDropManagerHelper.GetInstance().Cancel();
        _groot.HideTooltips();
        _groot.HidePopup();
        _groot.CloseModalWait();
        _groot.CloseAllWindows();

        var i = _groot.NumChildren() - 1;
        while (i >= 0)
        {
            var child = _groot.GetChildAt(i);
            if (child == _view)
            {
                i--;
                continue;
            }
            if (child.GetSortingOrder() >= 100000)
            {
                i--;
                continue;
            }
            if (child.GetSortingOrder() >= 50000)
                _groot.RemoveChildAt(i);
            else
                i--;
        }
    }

    protected void RequestSceneChange(string scenePath)
    {
        if (!IsSceneActive())
            return;
        _pendingScenePath = scenePath;
        CallDeferred(MethodName.DeferredFinishSceneChange);
    }

    private void DeferredFinishSceneChange()
    {
        if (string.IsNullOrEmpty(_pendingScenePath) || !IsSceneActive())
        {
            _pendingScenePath = "";
            return;
        }
        CleanupGrootOverlays();
        CallDeferred(MethodName.DeferredDetachGroot);
    }

    private void DeferredDetachGroot()
    {
        if (string.IsNullOrEmpty(_pendingScenePath) || !IsSceneActive())
        {
            _pendingScenePath = "";
            return;
        }
        GRoot.Cleanup();
        _view = null;
        _groot = null;
        CallDeferred(MethodName.DeferredChangeScene);
    }

    private void DeferredChangeScene()
    {
        var scenePath = _pendingScenePath;
        _pendingScenePath = "";
        if (string.IsNullOrEmpty(scenePath))
            return;
        GetEngineTree()?.CallDeferred(SceneTree.MethodName.ChangeSceneToFile, scenePath);
    }

    protected virtual void OnClose()
    {
        RequestSceneChange(MainMenuScenePath);
    }
}

File: examples/ts/Scripts/DemoSceneBase.ts
/// <reference path="../fairygui.d.ts" />

import "./fgui-globals";
import { Callable, Engine, Node, SceneTree } from "godot";

export type GodotNode = Node & {
    isInsideTree(): boolean;
    getTree(): SceneTree;
    callDeferred(method: string, ...args: unknown[]): void;
};

export class DemoSceneBase extends Node {
    protected mainMenuScenePath = "res://ts/Scenes/MainMenu.tscn";

    protected _groot: GRoot | null = null;
    protected _view: GComponent | null = null;

    private _sceneActive = false;
    private _pendingScenePath: string | null = null;

    _ready(): void {
        this._sceneActive = true;
        this._registerDefaultFonts();
        this._registerDefaultScrollBars();
        if (GRoot.getInstance() != null) {
            GRoot.cleanup();
        }
        (this as unknown as GodotNode).callDeferred("_delayedInit");
    }

    _exit_tree(): void {
        this._view = null;
        this._groot = null;
        this._sceneActive = false;
        this._pendingScenePath = null;
    }

    protected isSceneActive(): boolean {
        return this._sceneActive;
    }

    protected isUiActive(): boolean {
        const self = this as unknown as GodotNode;
        if (!this._sceneActive || !self.isInsideTree()) {
            return false;
        }
        if (this._groot == null || this._view == null) {
            return false;
        }
        if (GRoot.getInstance() == null) {
            return false;
        }
        return GRoot.getInstance() != null;
    }

    /** 不依赖场景节点，切场景后仍可用。 */
    protected getEngineTree(): SceneTree | null {
        const loop = Engine.getMainLoop();
        if (loop == null) {
            return null;
        }
        return loop as SceneTree;
    }

    /** GodotJS 在节点已离树时调用 getTree() 会报错，须先 isInsideTree()。 */
    protected safeGetTree(): SceneTree | null {
        const self = this as unknown as GodotNode;
        if (!this._sceneActive || !self.isInsideTree()) {
            return null;
        }
        return self.getTree();
    }

    _delayedInit(): void {
        if (!this.isSceneActive()) {
            return;
        }
        const tree = this.safeGetTree();
        if (tree == null) {
            return;
        }
        if (GRoot.getInstance() == null) {
            GRoot.create(this as unknown as Node);
        }
        this._deferredAttachToGroot();
    }

    _deferredAttachToGroot(): void {
        if (!this.isSceneActive()) {
            return;
        }
        this._groot = GRoot.getInstance();
        this._prepareGrootForScene();
        (this as unknown as GodotNode).callDeferred("_finishContinueInit");
    }

    _finishContinueInit(): void {
        if (!this.isSceneActive() || this._groot == null) {
            return;
        }
        void this.ContinueInit();
        this._addCloseButton();
    }

    protected _prepareGrootForScene(): void {
        if (this._groot == null) {
            return;
        }
        if (DragDropManagerHelper.getInstance().isDragging()) {
            DragDropManagerHelper.getInstance().cancel();
        }
        this._groot.hideTooltips();
        this._groot.hidePopup();
        this._groot.closeModalWait();
        this._groot.closeAllWindows();
        this._groot.removeChildren();
    }

    protected _registerDefaultFonts(): void {
        const fontPath = "res://Resources/fonts/NotoSansSC-Regular.ttf";
        const ui = UIConfigHelper.getInstance()!;
        ui.registerFont("default", fontPath);
        ui.registerFont("en", fontPath);
        ui.registerFont("微软雅黑", fontPath);
        ui.defaultFont = "default";
    }

    /** ScrollBar 组件在 Basics 包里；各 demo 须先注册 URL 并加载包，否则 ScrollPane 创建不出滚动条。 */
    protected _registerDefaultScrollBars(): void {
        UIPackage.addPackage("res://Resources/UI/Basics");
        const ui = UIConfigHelper.getInstance()!;
        ui.verticalScrollBar = "ui://Basics/ScrollBar_VT";
        ui.horizontalScrollBar = "ui://Basics/ScrollBar_HZ";
    }

    protected ContinueInit(): void | Promise<void> {
        // Override in subclasses.
    }

    protected async waitSeconds(seconds: number): Promise<void> {
        if (!this.isSceneActive()) {
            return;
        }
        const tree = this.safeGetTree() ?? this.getEngineTree();
        if (tree == null) {
            return;
        }
        const timer = tree.createTimer(seconds) as { timeout: { asPromise(): Promise<void> } };
        await timer.timeout.asPromise();
    }

    protected _addCloseButton(): void {
        if (!this._groot) {
            return;
        }
        const closeBtn = UIPackage.createObject("MainMenu", "CloseButton");
        if (closeBtn == null) {
            return;
        }
        closeBtn.setPosition(
            this._groot.getWidth() - closeBtn.getWidth() - 10,
            this._groot.getHeight() - closeBtn.getHeight() - 10,
        );
        closeBtn.addRelation(this._groot, FguiRelationType.RightRight, false);
        closeBtn.addRelation(this._groot, FguiRelationType.BottomBottom, false);
        closeBtn.setSortingOrder(100000);
        closeBtn.addClickListener(
            Callable.create(() => {
                (this as unknown as GodotNode).callDeferred("_onClose");
            }),
        );
        this._groot.addChild(closeBtn);
    }

    protected _cleanupGrootOverlays(): void {
        if (this._groot == null) {
            return;
        }
        if (DragDropManagerHelper.getInstance().isDragging()) {
            DragDropManagerHelper.getInstance().cancel();
        }
        this._groot.hideTooltips();
        this._groot.hidePopup();
        this._groot.closeModalWait();
        this._groot.closeAllWindows();

        let i = this._groot.numChildren() - 1;
        while (i >= 0) {
            const child = this._groot.getChildAt(i);
            if (child === this._view) {
                i -= 1;
                continue;
            }
            if (child.getSortingOrder() >= 100000) {
                i -= 1;
                continue;
            }
            if (child.getSortingOrder() >= 50000) {
                this._groot.removeChildAt(i);
            } else {
                i -= 1;
            }
        }
    }

    /** 延迟清空 GRoot 并切场景，分帧 detach 避免 FUIContainer 信号断开警告。 */
    protected _requestSceneChange(scenePath: string): void {
        if (!this.isSceneActive()) {
            return;
        }
        this._pendingScenePath = scenePath;
        (this as unknown as GodotNode).callDeferred("_deferredFinishSceneChange");
    }

    _deferredFinishSceneChange(): void {
        if (this._pendingScenePath == null || !this.isSceneActive()) {
            this._pendingScenePath = null;
            return;
        }
        this._cleanupGrootOverlays();
        (this as unknown as GodotNode).callDeferred("_deferredDetachGroot");
    }

    _deferredDetachGroot(): void {
        if (this._pendingScenePath == null || !this.isSceneActive()) {
            this._pendingScenePath = null;
            return;
        }
        GRoot.cleanup();
        this._view = null;
        this._groot = null;
        (this as unknown as GodotNode).callDeferred("_deferredChangeScene");
    }

    _deferredChangeScene(): void {
        const scenePath = this._pendingScenePath;
        this._pendingScenePath = null;
        if (scenePath == null) {
            return;
        }
        this.getEngineTree()?.callDeferred("change_scene_to_file", scenePath);
    }

    _onClose(): void {
        this._requestSceneChange(this.mainMenuScenePath);
    }
}
